### PR TITLE
New utils module providing cached_property.

### PR DESCRIPTION
**Pull request by Piezoid**
More transparent cache on python classes. 

### cached_property(member function)

Like @property on a member function, but cache the calculation in `self.__dict__[function name]`.
The function is called only once since the cache stored as an instance attribute override the property residing in the class attributes. Following accesses cost no more than standard Python attribute access.
If the instance attribute is deleted the next access will re-evaluate the function.
Source: https://blog.ionelmc.ro/2014/11/04/an-interesting-python-descriptor-quirk/
usage:
```python
class Shape(object):

    @cached_property
    def area(self):
        # compute value
        return value
```

## HaveCachedProperties
Used as a base class it provides cache management for classes using cached_property.
It's not mandatory for using cached_property, but it facilitate the access of  all cached attributes uder the `.cache` property

```python
class TestProperty(HaveCachedProperties):
    @cached_property
    def test_property(self):
        return "Ok"

t = TestProperty()
assert t.cache == {}

t.test_property # compute the poperty and cache it
assert t.cache == {'test_property': 'Ok'}

t.reset_cache() # equivalent del t.cache, t.cache  = {}
```

### CODE REVIEW

File: askomics/libaskomics/utils.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    def __get__(self, obj, cls):
+        value = obj.__dict__[self.func.__name__] = self.func(obj)
+        return value
+
+
+class HaveCachedProperties(object):
+    """Provide cache management for classes using cached_property."""
+    @classmethod
+    def getCachedProperties(cls):
```
Comment (abretaud) [inline askomics/libaskomics/utils.py]: Any reason to use camelCase here?

File: askomics/test/source_file_test.py [hunk 2]
```diff
@@ -2,7 +2,7 @@
 import unittest
 import json
 import tempfile, shutil
-
+print('found')
```
Comment (abretaud) [inline askomics/test/source_file_test.py]: I guess this should be removed?

File: askomics/libaskomics/utils.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+"""Some python magic to provide utility Bases, decorators, utility functions, etc. """
+
+import sys
+from pprint import pformat
+
+__all__ = ['pformatGenericObject',
+           'cached_property', 'HaveCachedProperties']
+
+
+def pformatGenericObject(obj):
```
Comment (abretaud) [inline askomics/libaskomics/utils.py]: Why camelCase here too?

File: askomics/libaskomics/source_file/SourceFile.py [hunk 4]
```diff
@@ -4,10 +4,17 @@
 import re
 import logging
 import csv
+from collections import defaultdict
 from itertools import count
 import os.path
 
-class SourceFile(object):
+
+from ..utils import cached_property, HaveCachedProperties, pformatGenericObject
```
Comment (abretaud) [inline askomics/libaskomics/source_file/SourceFile.py]: I would prefer to import from "askomics.libaskomics.utils"

File: askomics/libaskomics/source_file/SourceFile.py [hunk 5]
```diff
@@ -233,15 +225,47 @@ def get_domain_knowledge(self):
         # FIXME what do we do if self.category_values is empty or if it contains data from preview only?
         # FIXME maybe we could allow to get domain only when full turtle has been generated?
 
+
+    @cached_property
+    def category_values(self):
```
Comment (abretaud) [inline askomics/libaskomics/source_file/SourceFile.py]: Can you add docstring here?

File: askomics/libaskomics/source_file/SourceFile.py [hunk 6]
```diff
@@ -251,36 +275,37 @@ def get_turtle(self, preview_only=False):
...
                     if current_type == 'category':
                         # This is a category, keep track of allowed values for this column
-                        self.category_values.setdefault(header, set()).add(row[i])
+                        self.category_values[header].add(row[i])
 
                     # Create link to value
                     ttl += indent + " :has_" + header + " " + self.delims[current_type][0] + row[i] + self.delims[current_type][1] + " ;\n"
 
-                ttl = attribute_code[:-2] + ".\n"
+                ttl = attribute_code[:-2] + ".\n" #FIXME: Something is missing here
```
Comment (abretaud) [inline askomics/libaskomics/source_file/SourceFile.py]: What is missing?

**Comment by abretaud** [comment]
It looks good, thanks!
I'll pull it once you have checked the notes I added

File: askomics/libaskomics/source_file/SourceFile.py (hunk 6 above)
Comment (Piezoid) [inline askomics/libaskomics/source_file/SourceFile.py]: `attribute_code` is never defined. I guess it's from the old source code.

Comment (abretaud) [inline askomics/libaskomics/source_file/SourceFile.py]: Ah ok, looking at the old code, it should be "ttl[:-2]"
(not sure if it's still useful though !?)

**Pull request merged**